course_contents/2_intro_to_python/notes/03. Comprehensions.md

```diff
@@ -0,0 +1,99 @@
+### The list comprehension
+- the list comprehension allows us to create a list without manually appending objects or running loops
+- the general expression for a list comprehension is:
+
+```py
+new_collection = [element for element in collection] 
+```
+
+- the `element` variable will take the value of each element from the `collection` variable and be appended to the `new_collection`
+
+An example of a list comprehension:
+
+```py
+numbers = [i for i in range(5)]
+print(numbers)
+```
+
+#### List comprehension with conditionals
+- conditionals can be added to list comprehensions to make them more versatile
+
+An example of a list comprehension with a conditional:
+
+```py
+numbers = [i for i in range(5) if i % 2 == 0] #appending even numbers only
+print(numbers)
+```
+
+### Set comprehension
+- set comprehensions follow the same syntax as a list comprehension
+- the only difference is that curly brackets are used instead of square brackets
+
+An example of a set comprehension:
+
+```py
+collection = [1, 2, 2, 3, 4, 3, 5]
+numbers = {i for i in collection}
+print(numbers)
+```
+
+### Dictionary comprehension
+- dictionary comprehensions follow the next syntax:
+
+```py
+dictionary = {
+    collection_1[i] : collection_2[i]
+    for i in collection 
+}
+```
+
+- the collection_1[i] will be the ith element from the collection_1 variable and it will be the key
+- the collection_2[i] will be the ith element from the collection_2 variable and it will be the value
+
+An example of a dictionary comprehension:
+
+```py
+friends = ["Rolf", "Dane", "Josh"]
+numbers = [1,2,3]
+dictionary = {
+    friends[i] : numbers[i]
+    for i in range(3)
+}
+print(dictionary)
+```
+
+### Comprehensions with conditions
+- an if statement can be added to the comprehension to make it more versatile
+- if the conditional is correct, the element will be added to the list
+- the general expression for a list comprehension with a conditional is:
+
+```py
+new_collection = [element for element in collection if conditional]  
+```
+
+An example of a list comprehension with a condition:
+
+```py
+collection = [1, 2, 3, 4, 5]
+new_collection = [element for element in collection if element % 2 == 0] 
+print(new_collection)
+```
+
+### List slicing
+- list slicing is used to extract a part of the list
+- syntax for list slicing:
+
+```py
+example_list[start : stop : step]
+```
+
+- the last index that will be accessed is equal to `stop`-1
+- if `stop` is not defined the whole list after start will be sliced 
+- negative indexes can be used (-1 is the last element of the list)
+
+An example of list slicing:
+
+```py
+example_list = [1,2,3,4,5]
+print(example_list[1:4])
```

Comment:
Could you also show an example of negative slicing?

Reply:
I also thought of this, but then again, I went for shortness over explaining